### PR TITLE
feat(perp): MsgShiftPegMultiplier, MsgShiftSwapInvariant

### PR DESCRIPTION
**Pull request by Unique-Divine**
# Purpose

- Needed for testing the repeg service in the go-hm repo
- Relates to https://github.com/NibiruChain/nibiru/issues/1580

## Summary by CodeRabbit

- **New Features**
  - Introduced `MsgShiftPegMultiplier` to adjust peg multiplier parameters for perpetual contracts.
  - Added `MsgShiftSwapInvariant` for modifying swap invariant parameters in perpetual contracts.

- **Improvements**
  - Enhanced admin functions with permission checks for modifying AMM pool parameters.
  - Streamlined event emission for peg multiplier and swap invariant adjustments.

- **Bug Fixes**
  - Corrected error message from `ErrNegativeSwapInvariant` to `ErrNonPositiveSwapInvariant` for clarity.

- **Refactor**
  - Renamed functions related to price multiplier and swap invariant adjustments for better readability.
  - Removed deprecated functions `PegShift` and `DepthShift` from various modules.

- **Tests**
  - Updated and added new test cases to cover the new admin functionalities and message types.
  - Removed obsolete test functions related to removed peg and depth shift functionalities.

- **Documentation**
  - Updated CHANGELOG.md to reflect new features and changes in the perpetual contracts module.

### CODE REVIEW

File: x/perp/v2/keeper/admin.go [hunk 1]
```diff
@@ -174,3 +174,33 @@
...
+func (k admin) ShiftSwapInvariant(
+	ctx sdk.Context,
+	pair asset.Pair,
+	newSwapInvariant sdk.Dec,
+	sender sdk.AccAddress,
+) error {
+	if err := k.SudoKeeper.CheckPermissions(sender, ctx); err != nil {
+		return err
+	}
+	return k.UnsafeShiftSwapInvariant(ctx, pair, newSwapInvariant)
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/perp/v2/keeper/admin.go:205]: That's inside the call for UnsafeShiftSwapInvariant

File: x/perp/v2/keeper/amm_test.go [hunk 2]
```diff
@@ -185,11 +185,11 @@ func TestEditPriceMultiplerFail(t *testing.T) {
...
-	err = app.PerpKeeperV2.Admin.EditPriceMultiplier(ctx, asset.MustNewPair("luna:usdt"), sdk.NewDec(-1))
+	err = app.PerpKeeperV2.Admin.UnsafeShiftPegMultiplier(ctx, asset.MustNewPair("luna:usdt"), sdk.NewDec(-1))
 	require.ErrorContains(t, err, "market luna:usdt not found")
 
 	// Error because of invalid price multiplier
-	err = app.PerpKeeperV2.Admin.EditPriceMultiplier(ctx, pair, sdk.NewDec(-1))
+	err = app.PerpKeeperV2.Admin.UnsafeShiftPegMultiplier(ctx, pair, sdk.NewDec(-1))
 	require.ErrorIs(t, err, types.ErrNonPositivePegMultiplier)
 
 	// Add market activity
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/perp/v2/keeper/amm_test.go:197]: Fixed in b081818de00b4ab67da4b7ad2e217c53acddd047

File: x/perp/v2/keeper/admin.go [hunk 3]
```diff
@@ -174,3 +174,33 @@ func (k admin) UnsafeChangeCollateralDenom(
...
+) error {
+	if err := k.SudoKeeper.CheckPermissions(sender, ctx); err != nil {
+		return err
+	}
+	return k.UnsafeShiftPegMultiplier(ctx, pair, newPriceMultiplier)
+}
+
+// ShiftSwapInvariant: Edit the peg multiplier of an amm pool after making
+// sure there's enough money in the perp EF fund to pay for the repeg. These
+// funds get send to the vault to pay for trader's new net margin.
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/perp/v2/keeper/admin.go]: Fixed in b081818de00b4ab67da4b7ad2e217c53acddd047

File: x/perp/v2/keeper/amm_test.go [hunk 4]
```diff
@@ -21,7 +21,7 @@ import (
 	types "github.com/NibiruChain/nibiru/x/perp/v2/types"
 )
 
-func TestEditPriceMultipler(t *testing.T) {
+func TestUnsafeShiftPegMultiplier(t *testing.T) {
 	pair := asset.Registry.Pair(denoms.BTC, denoms.NUSD)
 
 	tests := TestCases{
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/perp/v2/keeper/amm_test.go:27]: Addressed 1b5c83654bfaeaa06b88b6971f8a7631c712248b

File: x/perp/v2/keeper/amm_test.go [hunk 5]
```diff
@@ -277,19 +281,19 @@
...
 
 	// Error because no money in perp ef fund
-	err = app.PerpKeeperV2.Admin.EditSwapInvariant(ctx, pair, sdk.NewDec(2_000_000))
+	err = app.PerpKeeperV2.Admin.UnsafeShiftSwapInvariant(ctx, pair, sdk.NewDec(2_000_000))
 	require.ErrorContains(t, err, "not enough fund in perp ef to pay for repeg")
 
 	// Fail at validate
-	err = app.PerpKeeperV2.Admin.EditSwapInvariant(ctx, pair, sdk.NewDec(0))
+	err = app.PerpKeeperV2.Admin.UnsafeShiftSwapInvariant(ctx, pair, sdk.NewDec(0))
 	require.ErrorContains(t, err, "swap multiplier must be > 0")
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/perp/v2/keeper/amm_test.go]: Fixed by 2f0a1256cc3f37fa79d3bf382f5cbb1848eb6d0b

File: x/perp/v2/keeper/msg_server.go [hunk 6]
```diff
@@ -189,3 +188,27 @@ func (m msgServer) WithdrawEpochRebates(ctx context.Context, msg *types.MsgWithd
...
 	}, nil
 }
+
+// ShiftPegMultiplier: gRPC tx msg for changing a market's peg multiplier.
+// [Admin] Only callable by sudoers.
+func (m msgServer) ShiftPegMultiplier(
+	goCtx context.Context, msg *types.MsgShiftPegMultiplier,
+) (*types.MsgShiftPegMultiplierResponse, error) {
+	// The `msg` here is checked with ValidateBasic before this fn is called.
+	sender, _ := sdk.AccAddressFromBech32(msg.Sender)
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/perp/v2/keeper/msg_server.go:198]: Covered by `ValidateBasic`

File: x/perp/v2/keeper/msg_server.go [hunk 7]
```diff
@@ -189,3 +188,27 @@
...
+	return &types.MsgShiftPegMultiplierResponse{}, err
+}
+
+// ShiftSwapInvariant: gRPC tx msg for changing a market's swap invariant.
+// [Admin] Only callable by sudoers.
+func (m msgServer) ShiftSwapInvariant(
+	goCtx context.Context, msg *types.MsgShiftSwapInvariant,
+) (*types.MsgShiftSwapInvariantResponse, error) {
+	// The `msg` here is checked with ValidateBasic before this fn is called.
+	sender, _ := sdk.AccAddressFromBech32(msg.Sender)
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/perp/v2/keeper/msg_server.go:210]: Covered by `ValidateBasic`

File: x/perp/v2/integration/action/market.go [hunk 8]
```diff
@@ -140,7 +140,7 @@ type editPriceMultiplier struct {
 }
 
 func (e editPriceMultiplier) Do(app *app.NibiruApp, ctx sdk.Context) (sdk.Context, error, bool) {
-	err := app.PerpKeeperV2.EditPriceMultiplier(ctx, e.pair, e.newValue)
+	err := app.PerpKeeperV2.UnsafeShiftPegMultiplier(ctx, e.pair, e.newValue)
 	return ctx, err, true
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/perp/v2/integration/action/market.go:144]: They are two different things. The swap invariant is specified with an int, while the peg multiplier is not.

File: x/perp/v2/keeper/amm_test.go [hunk 9]
```diff
@@ -368,7 +372,7 @@ func TestEditSwapInvariant(t *testing.T) {
 				FundModule(types.PerpEFModuleAccount, sdk.NewCoins(sdk.NewCoin(types.TestingCollateralDenomNUSD, sdk.NewInt(1e6)))),
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/perp/v2/keeper/amm_test.go:372]: Fixed in 552a2d5da3923c7eead2a0d8b656c22cd97bf253

File: x/perp/v2/keeper/amm_test.go [hunk 10]
```diff
@@ -388,7 +392,7 @@ func TestEditSwapInvariant(t *testing.T) {
...
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/perp/v2/keeper/amm_test.go:392]: Fixed in 552a2d5da3923c7eead2a0d8b656c22cd97bf253

File: x/perp/v2/keeper/amm_test.go [hunk 11]
```diff
@@ -348,7 +352,7 @@ func TestEditSwapInvariant(t *testing.T) {
...
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/perp/v2/keeper/amm_test.go:352]: Fixed in 552a2d5da3923c7eead2a0d8b656c22cd97bf253

File: x/perp/v2/keeper/amm_test.go [hunk 12]
```diff
@@ -328,7 +332,7 @@ func TestEditSwapInvariant(t *testing.T) {
...
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/perp/v2/keeper/amm_test.go:332]: Fixed in 552a2d5da3923c7eead2a0d8b656c22cd97bf253

File: x/perp/v2/keeper/amm_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/perp/v2/keeper/amm_test.go:27]: Fixed in 552a2d5da3923c7eead2a0d8b656c22cd97bf253

File: x/perp/v2/keeper/amm.go [hunk 13]
```diff
@@ -78,7 +82,9 @@ func (k Keeper) EditSwapInvariant(ctx sdk.Context, pair asset.Pair, newSwapInvar
 	return nil
 }
 
-func (k Keeper) handleMarketUpdateCost(ctx sdk.Context, pair asset.Pair, costAmt sdkmath.Int) (err error) {
+func (k Keeper) handleMarketUpdateCost(
+	ctx sdk.Context, pair asset.Pair, costAmt sdkmath.Int,
+) (err error) {
 	collateral, err := k.Collateral.Get(ctx)
 	if err != nil {
 		return err
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/perp/v2/keeper/amm.go]: Addressed by emitting events to express the cost paid

File: x/perp/v2/keeper/admin.go [hunk 14]
```diff
@@ -174,3 +174,33 @@ func (k admin) UnsafeChangeCollateralDenom(
...
+func (k admin) ShiftPegMultiplier(
+	ctx sdk.Context,
+	pair asset.Pair,
+	newPriceMultiplier sdk.Dec,
+	sender sdk.AccAddress,
+) error {
+	if err := k.SudoKeeper.CheckPermissions(sender, ctx); err != nil {
+		return err
+	}
+	return k.UnsafeShiftPegMultiplier(ctx, pair, newPriceMultiplier)
```
Comment (jgimeno) [inline x/perp/v2/keeper/admin.go:190]: I would do this function private as calling it from outside can make the application inconsistent. If we can avoid it that is good.

File: x/perp/v2/keeper/amm.go [hunk 15]
```diff
@@ -11,10 +11,11 @@ import (
...
 
-// EditPriceMultiplier edits the peg multiplier of an amm pool after making
-// sure there's enough money in the perp EF fund to pay for the repeg. These
-// funds get send to the vault to pay for trader's new net margin.
-func (k Keeper) EditPriceMultiplier(
+// UnsafeShiftPegMultiplier: [Without checking x/sudo permissions] Edits the peg
+// multiplier of an amm pool after making sure there's enough money in the perp
+// EF fund to pay for the repeg. These funds get send to the vault to pay for
+// trader's new net margin.
+func (k Keeper) UnsafeShiftPegMultiplier(
```
Comment (jgimeno) [inline x/perp/v2/keeper/amm.go:18]: commented before, I would try to make it private

File: x/perp/v2/keeper/admin.go (hunk 14 above)
Comment (Unique-Divine) [reply x/perp/v2/keeper/admin.go]: Done. As an even better step, I gave the sudo root address a convenience getter function inside `testapp.go` and deleted "Unsafe..." entirely for both messages

fe1fe03cb9cf99369486d7455c06a3d43c5eef25

File: x/perp/v2/keeper/amm.go (hunk 15 above)
Comment (Unique-Divine) [reply x/perp/v2/keeper/amm.go]: [Ditto](https://github.com/NibiruChain/nibiru/pull/1680#discussion_r1408596611)

**Pull request merged**